### ISSUE TITLE
getHistory generates very slow flux statement for InfluxDB 2.x

### ISSUE THREAD

**Issue by klein0r** [ioBroker/ioBroker.influxdb#328]
**Describe the bug**  

A user reported a very slow query, which hits the default 30 second timeout:

```
2023-03-06 10:40:04.668	warn Error in query "from(bucket: "iobroker_data") \|> range(start: 2021-03-05T23:00:00.000Z, stop: 2023-03-05T22:59:59.999Z) \|> filter(fn: (r) => r["_measurement"] == "0_userdata.0.Photovoltaik.DC-Leistung_gesamt") \|> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value") \|> group() \|> sort(columns: ["_time"], desc: true) \|> limit(n: 1)": RequestTimedOutError: Request timed out
```

```
from(bucket: "iobroker_data") 
    |> range(start: 2021-03-05T23:00:00.000Z, stop: 2023-03-05T22:59:59.999Z) 
    |> filter(fn: (r) => r["_measurement"] == "0_userdata.0.Photovoltaik.DC-Leistung_gesamt") 
    |> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value") 
    |> group() |> sort(columns: ["_time"], desc: true) 
    |> limit(n: 1)
```

It turns out, that the statement needs ~53 seconds to complete and it was executed by [flot](https://github.com/ioBroker/ioBroker.flot). Maybe [here](https://github.com/ioBroker/ioBroker.flot/blob/f112917fab740b1ac72739496395a682d4d6d41e/src/js/data.js#L393-L399)?

![1678110604134-8003f3dd-a1f6-461c-8b25-ee45fd9b5b42-image](https://user-images.githubusercontent.com/1959049/223501763-234b7298-1d67-4394-bc1a-509e0412c456.png)

Obviously the time range is very large (2 years) and there must be a lot of data in the database. Nevertheless, the generated statement is "not that efficient". Everything in the time range is selected (and reversed), just to get the first value. That should be equal to ``last()``: https://docs.influxdata.com/flux/v0.x/stdlib/universe/last/

Is this the reason?

https://github.com/ioBroker/ioBroker.influxdb/blob/1a3998a992d87e91936f3b0f339a47f751237527/main.js#L2422-L2436

or this?

https://github.com/ioBroker/ioBroker.influxdb/blob/1a3998a992d87e91936f3b0f339a47f751237527/main.js#L2364-L2372

The question is: Is this an issue with the influxdb adapter for an optimization of the statement, or are too many information requested by flot and the request needs to be updated?

More details in the [iobroker forums](https://forum.iobroker.net/topic/63613/warn-logeinträge-influxdb-adapter).

**Comment by Apollon77** [comment]
In fact a debug log of such a datapoint and a getHistory request would be very helpful.

In fact the queries were written with the "best guess" understanding about fluxQL and I really think that they are not optimized for selection performance ;-)

If "last" would replace such stuff in the query then it could be an idea to adjust this and yes I think that this is a topic of influxdb adapter to optimize. But because of limited knowledge of fluxQL someone with a running InfluxDB v2 should do and test the change to make sure it really produces the same output

**Comment by mango1402** [comment]
Sorr, kenn mich nix aus mit GIT und PR...
Anbei mein adaptiertes file.

[main.js.txt](https://github.com/ioBroker/ioBroker.influxdb/files/10948837/main.js.txt)

**Comment by mango1402** [comment]
> @mango1402 Great work, thanks for analyzing the issue.

would be great if somebody approves my 'bug-fix' and provides a new release so that all users get a benefit......

**Comment by Apollon77** [comment]
Yes I will do that soon

**Comment by Apollon77** [comment]
Yes but this is a different issue ... check the open issues, should be there also already

**Comment by Apollon77** [comment]
I think it is not yet released, or? So only GitHub version has it ... I hope I come to it soon, sorry

**Comment by klein0r** [comment]
Ah okay, I thought it's implemented since 3.1.4

https://github.com/ioBroker/ioBroker.influxdb/blob/224a6e32ac35a6441414e8a997eb170c003a825e/README.md?plain=1#L473-L474

**Comment by Marc-Berg** [comment]
Now that version 4.0.2 has been released, can this issue be closed, I think?
